walnut/validate/validate.go

```diff
@@ -0,0 +1,100 @@
+/*
+ * === This file is part of ALICE O² ===
+ *
+ * Copyright 2020 CERN and copyright holders of ALICE O².
+ * Author: Ayaan Zaidi <azaidi@cern.ch>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * In applying this license CERN does not waive the privileges and
+ * immunities granted to it by virtue of its status as an
+ * Intergovernmental Organization or submit itself to any jurisdiction.
+ */
+
+package validate
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/AliceO2Group/Control/walnut/app"
+
+	"github.com/xeipuuv/gojsonschema"
+	"gopkg.in/yaml.v2"
+)
+
+// Template accepts a filename and format then validate against the schema specified (either workflow or task)
+func Template(filename string, format string) {
+
+	rawYAML, err := ioutil.ReadFile(filename) // import YAML file
+	if err != nil {
+		fmt.Printf("ReadFile failed: %v", err)
+		os.Exit(1)
+	}
+
+	var dataFromYAML interface{}
+	if err := yaml.Unmarshal([]byte(rawYAML), &dataFromYAML); err != nil {
+		fmt.Printf("Unmarshaling YAML failed: %v", err)
+		os.Exit(1)
+	}
+
+	dataFromYAML = convert(dataFromYAML)
+
+	var schema string
+	switch format {
+	case "task":
+		schema = app.TaskSchema
+	case "workflow":
+		schema = app.WorkflowSchema
+	default:
+		fmt.Print("ERROR: Invalid format specified.")
+		os.Exit(1)
+	}
+
+	schemaLoader := gojsonschema.NewStringLoader(schema)     // load schema
+	documentLoader := gojsonschema.NewGoLoader(dataFromYAML) // load empty interface
+
+	result, err := gojsonschema.Validate(schemaLoader, documentLoader)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	if result.Valid() {
+		fmt.Printf("\nSUCCESS! File: %s is valid against %s schema.", filename, format)
+		os.Exit(0)
+	} else {
+		err := "Schema validation failed."
+		fmt.Printf("ERROR: %v", err)
+		os.Exit(1)
+	}
+}
+
+// convert takes a interface{} as input and recursively converts all child
+// map[interface{}]interface{} to map[string]interface{}
+func convert(i interface{}) interface{} {
```

Comment:
This approach is ok.

I suspect the difficulty in unmarshaling to `map[string]interface{}` derives from the fact that YAML allows complex non-string keys. An alternative approach to yours would be to define a `type map[string]interface{} MyMapType` and implement `func (m *MyMapType) UnmarshalYAML(unmarshal func(interface{}) error) (err error)`. By handling the string keys in this function you make it possible to unmarshal directly to `MyMapType`. Try if you like, or we can leave it as it is (I don't expect a performance difference in our use case).

Reply:
I'm not sure I completely understand this approach. It does seem like a cleaner solution, let me know if you think it's worth exploring. I'm fine with it either way.

Reply:
It is not critical in this specific case, but I expect understanding and using the `yaml.Unmarshaler`/`yaml.Marshaler` interfaces is stuff you'll need for the workflow conversion effort that follows, where non-trivial YAML scenarios make your current approach impractical.
I advise checking out https://godoc.org/gopkg.in/yaml.v2 as well as the files in https://github.com/AliceO2Group/Control/tree/master/core/workflow and https://github.com/AliceO2Group/Control/tree/master/core/task which use `yaml.Unmarshaler` extensively to control YAML-to-memory deserialization.

For example, if we consider `channel.go`, implementing the `UnmarshalYAML` function allows us to deserialize those `sndBufSize`, `rateLogging` etc. into an auxiliary type (`_channel`) which directly mimics the YAML structure. We then do some casts and fill in some defaults, and store this in the actual target object whose type doesn't directly match YAML without this special constructor.